test/integration/helpers_test.go

```diff
@@ -0,0 +1,168 @@
+// and finally deploys it on the Kubernetes cluster by means of the gateway client given as arg.
+func DeployGateway(ctx context.Context, client *gatewayclient.Clientset, namespace, gatewayClassName string, opts ...func(*gatewayv1alpha2.Gateway)) (*gatewayv1alpha2.Gateway, error) {
+	var err error
+
+	// create a default gateway with a listener set to port 80 for HTTP traffic
+	gw := &gatewayv1alpha2.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: defaultGatewayName,
+			Annotations: map[string]string{
+				unmanagedAnnotation: "true", // trigger the unmanaged gateway mode
+			},
+		},
+		Spec: gatewayv1alpha2.GatewaySpec{
+			GatewayClassName: gatewayv1alpha2.ObjectName(gatewayClassName),
+			Listeners: []gatewayv1alpha2.Listener{{
+				Name:     "http",
+				Protocol: gatewayv1alpha2.HTTPProtocolType,
+				Port:     gatewayv1alpha2.PortNumber(ktfkong.DefaultProxyTCPServicePort),
+			}},
+		},
+	}
+
+	// call all the modifiers eventually passed as args
+	for _, opt := range opts {
+		opt(gw)
+	}
+
+	if gw, err = client.GatewayV1alpha2().Gateways(namespace).Create(ctx, gw, metav1.CreateOptions{}); err != nil {
+		return nil, err
+	}
+
+	return gw, nil
+}
+
+// gatewayHealthCheck checks the gateway has been scheduled by KIC. This function is inspired by
+// assert.eventually (https://github.com/stretchr/testify/blob/v1.7.5/assert/assertions.go#L1669-L1700)
+// and patched with custom behavior to determine the health of the gateway and to return an error
+// instead of failing (at the time of its call, we don't have any test instance to make fail yet).
+func gatewayHealthCheck(ctx context.Context, client *gatewayclient.Clientset, gatewayName, namespace string) error {
+	ch := make(chan bool, 1)
+
+	timer := time.NewTimer(gatewayWaitTimeToVerifyScheduling)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for tick := ticker.C; ; {
+		select {
+		case <-timer.C:
+			return fmt.Errorf("the gateway scheduling %s/%s has never occurred", namespace, defaultGatewayName)
+		case <-tick:
+			tick = nil
+			go func() {
+				ch <- func() bool {
+					gw, err := client.GatewayV1alpha2().Gateways(namespace).Get(ctx, gatewayName, metav1.GetOptions{})
+					exitOnErr(err)
+					for _, cond := range gw.Status.Conditions {
+						if cond.Reason == string(gatewayv1alpha2.GatewayReasonReady) {
+							return true
+						}
+					}
+					return false
+				}()
+			}()
+		case v := <-ch:
+			if v {
+				return nil
+			}
+			tick = ticker.C
+		}
+	}
+}
+
+// eventuallyGatewayIsLinked checks if the specific Route (HTTP, TCP, TLS, or UDP)
+// is correctly linked to (or unlinked from) a supported gateway. In order to assert
+// that the route must be linked to the gateway, or unlinked from the gateway, the
+// ensureLinked boolean arg must be set accordingly.
+func eventuallyGatewayIsLinked(t *testing.T, c *gatewayclient.Clientset, ensureLinked bool, protocolType gatewayv1alpha2.ProtocolType, namespace, name string) {
```

Comment:
Yes, the original intention was to reduce as much as possible the code redundancy. Since eventuallyGatewayIsLinked and eventuallyGatewayIsUnlinked were the same functions apart from the return value, I thought that it may be worth adding that flag. Nevertheless, I agree that the code's clarity may be negatively affected. I'll introduce those wrappers as you suggested.